In [14]:
import pandas as pd

# Работа с Pandas

Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [15]:
df = pd.read_csv(r'E:/1T/lesson_1_5/googleplaystore.csv')
df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type  Price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

Отобразите количество строк и столбцов в DataFrame

In [16]:
df.shape

(10841, 13)

# PRO часть (Не обязательна к выполнению, но желательна :))


Измените названия всех столбцов приведя их к нижнему регистру

In [17]:
lower_column_names = []
for word in df.columns:
    lower_column_names.append(word.replace(" ", "_").lower()) # удаляю пробелы и привожу к нижнему регистру
df.columns = lower_column_names
print(df.columns)

Index(['app', 'category', 'rating', 'reviews', 'size', 'installs', 'type',
       'price', 'content_rating', 'genres', 'last_updated', 'current_ver',
       'android_ver'],
      dtype='object')


Выведите основные статистики по категориальным столбцам. Напишите выводы о них 

In [18]:
df.describe(include=object)

app category reviews                size    installs   type  \
count    10841    10841   10841               10841       10841  10840   
unique    9660       34    6002                 462          22      3   
top     ROBLOX   FAMILY       0  Varies with device  1,000,000+   Free   
freq         9     1972     596                1695        1579  10039   

       content_rating genres    last_updated         current_ver android_ver  
count           10840  10841           10841               10833       10838  
unique              6    120            1378                2832          33  
top          Everyone  Tools  August 3, 2018  Varies with device  4.1 and up  
freq             8714    842             326                1459        2451

**Выводы:**
- Из 10841 приложений 9660 имеют уникальные названия.
- Всего 34 уникальных категории приложений.
- Самое популярное приложение - ROBLOX, категория - FAMILY.
- Самый распространенный жанр - Tools.
- Большинство приложений не имеют reviews.
- В основном, приложения бесплатные и подходят для всех возрастных групп.
- Большинство приложений были обновлены в августе 2018 года.
- Большинство приложений требует версию операционной системы Android 4.1 и выше.

Выведите на экран часть датафрейма, который соответствует условию: `приложение из категории искусства и дизайна` И `с рейтингом больше 4.5`

In [19]:
df[(df.category == "ART_AND_DESIGN") & (df.rating > 4.5)]

app        category  \
2     U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN   
9                         Kids Paint Free - Drawing Fun  ART_AND_DESIGN   
13                                Mandala Coloring Book  ART_AND_DESIGN   
16         Photo Designer - Write your name with shapes  ART_AND_DESIGN   
19                                         ibis Paint X  ART_AND_DESIGN   
22              Superheroes Wallpapers | 4K Backgrounds  ART_AND_DESIGN   
24                          HD Mickey Minnie Wallpapers  ART_AND_DESIGN   
25                           Harley Quinn wallpapers HD  ART_AND_DESIGN   
26                        Colorfit - Drawing & Coloring  ART_AND_DESIGN   
34                                      I Creative Idea  ART_AND_DESIGN   
36       UNICORN - Color By Number & Pixel Art Coloring  ART_AND_DESIGN   
38                       PIP Camera - PIP Collage Maker  ART_AND_DESIGN   
45    Canva: Poster, banner, card maker & graphic de...  ART_AND_DESIGN   
46    Install images with music to make video withou...  ART_AND_DESIGN   
3982                                  Cardi B Wallpaper  ART_AND_DESIGN   
4749  X Launcher: With OS11 Style Theme & Control Ce...  ART_AND_DESIGN   
4755  X Launcher Pro: PhoneX Theme, OS11 Control Center  ART_AND_DESIGN   
4759  X Launcher Pro - IOS Style Theme & Control Center  ART_AND_DESIGN   
4764     X Launcher Prime: With OS Style Theme & No Ads  ART_AND_DESIGN   
5254                            AJ Styles HD Wallpapers  ART_AND_DESIGN   
6930               Fantasy theme dark bw black building  ART_AND_DESIGN   
8888            Spring flowers theme couleurs d t space  ART_AND_DESIGN   

      rating reviews  size     installs  type  price content_rating  \
2        4.7   87510  8.7M   5,000,000+  Free   0.00       Everyone   
9        4.7     121  3.1M      10,000+  Free   0.00       Everyone   
13       4.6    4326   21M     100,000+  Free   0.00       Everyone   
16       4.7    3632  5.5M     500,000+  Free   0.00       Everyone   
19       4.6  224399   31M  10,000,000+  Free   0.00       Everyone   
22       4.7    7699  4.2M     500,000+  Free   0.00   Everyone 10+   
24       4.7     118   23M      50,000+  Free   0.00       Everyone   
25       4.8     192  6.0M      10,000+  Free   0.00       Everyone   
26       4.7   20260   25M     500,000+  Free   0.00       Everyone   
34       4.7     353  4.2M      10,000+  Free   0.00           Teen   
36       4.7    8145   24M     500,000+  Free   0.00       Everyone   
38       4.7     158   11M      10,000+  Free   0.00       Everyone   
45       4.7  174531   24M  10,000,000+  Free   0.00       Everyone   
46       4.6    1070   26M     100,000+  Free   0.00       Everyone   
3982     4.8     253  3.7M      50,000+  Free   0.00       Everyone   
4749     4.7    5754  4.4M     100,000+  Free   0.00       Everyone   
4755     4.7     801  3.5M       5,000+  Paid   1.99       Everyone   
4759     4.8    1216  8.6M      10,000+  Paid   1.99       Everyone   
4764     4.7     149  3.5M       1,000+  Paid   1.99       Everyone   
5254     4.8      48   25M       5,000+  Free   0.00       Everyone   
6930     4.8      41  1.9M       5,000+  Free   0.00       Everyone   
8888     5.0       1  2.9M         100+  Free   0.00       Everyone   

                       genres       last_updated current_ver   android_ver  
2                Art & Design     August 1, 2018       1.2.4  4.0.3 and up  
9     Art & Design;Creativity       July 3, 2018         2.8  4.0.3 and up  
13               Art & Design      June 26, 2018       1.0.4    4.4 and up  
16               Art & Design      July 31, 2018         3.1    4.1 and up  
19               Art & Design      July 30, 2018       5.5.4    4.1 and up  
22               Art & Design      July 12, 2018     2.2.6.2  4.0.3 and up  
24               Art & Design       July 7, 2018       1.1.3    4.1 and up  
25               Art & Design     April 25, 2018         1.5    3.0 and up  

Посчитатайте какое количество игр содержится в рейтинге, используя параметр разбиения (bins) равным 100. Выведите на экран топ-10 игр по количеству в рейтинге. Напишите краткий вывод.
То есть сначала отфильтровываете лишь игры из всего датафрейма, после чего применяете value_counts к столбцу с рейтингом используя параметр, который указан выше. И выводите первые 10 строк отсортированных по убыванию.

In [20]:
df[df.category == "GAME"].rating.value_counts(bins=100).head(10)

(4.48, 4.52]    174
(4.36, 4.4]     166
(4.28, 4.32]    151
(4.56, 4.6]     121
(4.16, 4.2]     115
(4.08, 4.12]     72
(4.68, 4.72]     67
(3.96, 4.0]      55
(3.76, 3.8]      41
(3.88, 3.92]     35
Name: count, dtype: int64

Создайте функцию , которая будет применена к столбцу рейтинг методом `apply()`.
Функция должна принимать строку с рейтингом и возвращать:
 - High rating, если рейтинг больше или равен 4.5
 - Middle rating, если рейтинг между 3.8 и 4.5
 - Low rating, если рейтинг меньше 3.8 
Результат работы функции запишите в новый столбец с названием `categorical_rating`

In [21]:
def rating_category(row):
    if row >= 4.5:
        return "High rating"
    elif 3.8 <= row < 4.5:
        return "Middle rating"
    else:
        return "Low rating"
    
df["categorical_rating"] = df["rating"].apply(rating_category)
df.head()


app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type  price content_rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  
0  4.0.3 and up      Middle rating  
1  4.0.3 and up      Middle rating  
2  4.0.3 and up        High rating  
3    4.2 and up        High rating  
4    4.4 and up      Middle rating

Выведите кол-во уникальных значений в столбцах отсортированных по убыванию на экран:

In [22]:
df.nunique().sort_values(ascending=False)

app                   9660
reviews               6002
current_ver           2832
last_updated          1378
size                   462
genres                 120
price                   92
rating                  40
category                34
android_ver             33
installs                22
content_rating           6
type                     3
categorical_rating       3
dtype: int64

__Super-Hard задания__

Решите задачу, где мы создавали свою собственную функцию, но вместо функции `def` используйте `лямбда-функцию` внутри `apply()`

In [23]:
df['categorical_rating'] = df['rating'].apply(lambda x: 'High rating' if x >= 4.5 else ('Middle rating' if 3.8 <= x < 4.5 else 'Low rating'))
df.head()

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type  price content_rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  
0  4.0.3 and up      Middle rating  
1  4.0.3 and up      Middle rating  
2  4.0.3 and up        High rating  
3    4.2 and up        High rating  
4    4.4 and up      Middle rating

Изучите функцию `transform` в документации Pandas. Примените её к датафрейму, который будет заранее сгруппирован по столбцу `category`. С помощью фукнции `transform` найдите средний `rating` по каждой категории и запишите в столбец с названием `mean_cat_rating`. Кол-во строк в получившемся новом столбец должно быть таким же, как и в исходном датафрейме

In [24]:
df["mean_cat_rating"] = df.groupby("category")["rating"].transform("mean")
df.head()

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type  price content_rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  mean_cat_rating  
0  4.0.3 and up      Middle rating         4.358065  
1  4.0.3 and up      Middle rating         4.358065  
2  4.0.3 and up        High rating         4.358065  
3    4.2 and up        High rating         4.358065  
4    4.4 and up      Middle rating         4.358065

Создайте функцию , которая будет применена к столбцу `last updated` и возвращать из него число месяца, если число месяца вернуть из строки невозможно, то верните вместо этого строку с значением `miss_date`.\
Запишите полученный столбец в датафрейм , название столбцу дайте `day_of_update`.

Если вы столкнётесь с некоторой ошибкой , которую вам выдаст Pandas, то изучите конструкцию Python `try` \ `except`.\
Вот псевдокод , который поможет лучше разобраться с этой конструкцией:

In [32]:
def day_updated(row):
    try:
        return row.replace(",", "").split()[1]
    except:
        return "miss_date"

df["day_of_update"] = df["last_updated"].apply(day_updated)
df.sample(5)

app            category  \
7840                                   Helper for CS:GO               TOOLS   
9656                                            EO GSEA            BUSINESS   
2729  Coupons.com – Grocery Coupons & Cash Back Savings            SHOPPING   
5910                                    British Airways    TRAVEL_AND_LOCAL   
1414                      My Diet Diary Calorie Counter  HEALTH_AND_FITNESS   

      rating reviews                size    installs  type  price  \
7840     4.4     208                6.0M     10,000+  Free    0.0   
9656     NaN       1                 30M         10+  Free    0.0   
2729     4.2   24953  Varies with device  1,000,000+  Free    0.0   
5910     4.2   25726                 51M  1,000,000+  Free    0.0   
1414     4.1   18539                 57M  1,000,000+  Free    0.0   

     content_rating            genres     last_updated         current_ver  \
7840       Everyone             Tools    June 16, 2016               1.2.1   
9656       Everyone          Business  January 4, 2018               1.0.0   
2729       Everyone          Shopping    June 21, 2018  Varies with device   
5910       Everyone    Travel & Local    July 24, 2018                4.20   
1414       Everyone  Health & Fitness    July 30, 2018              1.11.4   

             android_ver categorical_rating  mean_cat_rating day_of_update  
7840          4.1 and up      Middle rating         4.047411            16  
9656          4.4 and up         Low rating         4.121452             4  
2729  Varies with device      Middle rating         4.259664            21  
5910          4.4 and up      Middle rating         4.109292            24  
1414        4.0.3 and up      Middle rating         4.277104            30